httpbakery/client.go

```diff
@@ -198,20 +180,30 @@ func (c *Client) dischargeAcquirer() DischargeAcquirer {
 	return c
 }
 
-// relativeURL returns newPath relative to an original URL.
-func relativeURL(base, new string) (*url.URL, error) {
-	if new == "" {
-		return nil, errgo.Newf("empty URL")
-	}
-	baseURL, err := url.Parse(base)
-	if err != nil {
-		return nil, errgo.Notef(err, "cannot parse URL")
-	}
-	newURL, err := url.Parse(new)
-	if err != nil {
-		return nil, errgo.Notef(err, "cannot parse URL")
-	}
-	return baseURL.ResolveReference(newURL), nil
+// Do is like DoWithContext, except the context is automatically derived.
+// If using go version 1.7 or later the context will be taken from the
+// given request, otherwise context.Background() will be used.
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
+	return c.do(contextFromRequest(req), req, nil)
+}
+
+// DoWithContext sends the given HTTP request and returns its response.
+// If the request fails with a discharge-required error, any required
+// discharge macaroons will be acquired, and the request will be repeated
+// with those attached.
+//
+// If the required discharges were refused by a third party, an error
+// with a *DischargeError cause will be returned.
+//
+// If interaction is required by the user, the client's WebPageVisitor,
+// or VisitWebPage function will be used to perform interaction. An error
+// with a *InteractionError cause will be returnedif this interaction
+// fails. See WebBrowserVisitor for a possible implementation of
+// WebPageVisitor.
+//
+// DoWithContext may add headers to req.Header.
+func (c *Client) DoWithContext(ctx context.Context, req *http.Request) (*http.Response, error) {
```

Comment:
I'm not sure about having another Do variant again now we've eliminated DoWithBody.. For example, what about DoWithContextAndCustomError?

How about just adding a context to Do and DoWithCustomError and don't bother to derive the context from the request?
At some point when we can use Go 1.7 we can break the API and take the context argument out and derive it from the request.

For now, I think all our pre-1.7 programs will need to pass in the request explicitly anyway.